snuba/utils/kafka/consumer.py

```diff
@@ -0,0 +1,30 @@
+from abc import ABC, abstractmethod
+from typing import Any, Callable, Generic, Optional, Sequence, TypeVar
+
+from confluent_kafka import TopicPartition
+
+
+TMessage = TypeVar("TMessage")
+
+
+class Consumer(Generic[TMessage], ABC):
```

Comment:
This is (roughly) the API that is used by the `BatchingKafkaConsumer`.

My thinking here: I would like to avoid introducing another way of running Kafka consumers to this project (and reimplementing things like commit batching, etc.)

To effectively support subscriptions, I think it will essentially require adding streaming support to the `BatchingKafkaConsumer` library, since I think it's important to be able to execute each `TaskSet` as soon as possible without waiting for a batch size limit or timeout. I do think it's useful (and important) to keep the same batch *commit* semantics, though (and configuration consistency.)

This will require some relatively substantial changes to the library (and ones that we've been talking about for a while with respect to streaming writes for memory usage.) My thinking is that we should vendor the `BatchingKafkaConsumer` into Snuba (at least temporarily) to avoid having to synchronize these changes across multiple projects.

Long-term, I'm not sure how we want to manage this dependency, as well as the `SynchronizedConsumer` dependency that will be introduced. Curious on everyone else's thoughts here.

Reply:
> Long-term, I'm not sure how we want to manage this dependency, as well as the `SynchronizedConsumer` dependency that will be introduced. Curious on everyone else's thoughts here.

I also think it's relatively important to have this code type checked. I haven't looked into how mypy works with third party modules that need to support Python 2 ([using comment-based syntax](https://mypy.readthedocs.io/en/latest/python2.html)) and 3.

Reply:
This would also tighten up the specification of the [`FakeKafkaConsumer`](https://github.com/getsentry/snuba/blob/d4fc8330b8e6abcd2d350f794933f6794975cf2f/tests/base.py#L76-L117) for testing, which I think is a good thing.

Reply:
This is happening in #485 (you're probably aware of this, mostly noting for posterity.)